File: python/src/dijkstra.py
"""Dijkstra's single-source shortest path algorithm for physical road networks.

Implements a binary min-heap (via Python's ``heapq`` module) variant of
Dijkstra's algorithm with lazy-deletion for stale priority queue entries.
Operates on pre-computed physical edge costs across three optimization modes:
FASTEST, CHEAPEST, and BALANCED.

Complexity: O((V + E) log V) time, O(V) space.
"""

import heapq
import time
from typing import Optional, Dict, Tuple, List

from models import RouteResult, Edge
from graph import Graph


def solve_dijkstra(graph: Graph, source: int, target: int, mode: str) -> Optional[RouteResult]:
    """Computes the cost-optimal route between two vertices using Dijkstra's algorithm.

    Executes a standard priority-queue-driven relaxation over the directed graph
    adjacency structure. Stale (superseded) heap entries are discarded efficiently
    via a lazy-deletion guard on popped cost versus recorded minimum.

    Path reconstruction proceeds via backtracking through the predecessor edge map,
    accumulating all physical metrics (distance, time, fuel, cost) along the route.

    Args:
        graph: Pre-built Graph instance with physics-evaluated edge costs.
        source: Integer ID of the origin vertex.
        target: Integer ID of the destination vertex.
        mode: Optimization objective. One of ``"FASTEST"``, ``"CHEAPEST"``, ``"BALANCED"``.

    Returns:
        A ``RouteResult`` containing the optimal path, aggregate physical metrics,
        and search statistics. Returns ``None`` if the target is unreachable from source.
    """
    start_time = time.perf_counter_ns()

    if source not in graph.nodes or target not in graph.nodes:
        return None

    if source == target:
        return RouteResult(
            source=source, destination=target, optimization=mode, algorithm="DIJKSTRA",
            route=[source], distance_km=0.0, travel_time_min=0.0, fuel_liters=0.0,
            fuel_cost=0.0, toll_cost=0.0, total_cost=0.0, nodes_explored=1,
            execution_time_us=int((time.perf_counter_ns() - start_time) / 1000)
        )

    # dist[v] = minimum accumulated cost to reach vertex v
    dist: Dict[int, float] = {}
    # prev_edge[v] = (predecessor_vertex, edge_traversed) for path reconstruction
    prev_edge: Dict[int, Tuple[int, Edge]] = {}

    # Priority queue entries: (accumulated_cost, vertex_id)
    pq = [(0.0, source)]
    dist[source] = 0.0
    nodes_explored = 0
    found = False

    while pq:
        d, u = heapq.heappop(pq)

        # Lazy deletion: discard entries superseded by a shorter path discovery
        if d > dist.get(u, float('inf')):
            continue

        nodes_explored += 1

        if u == target:
            found = True
            break

        for v, edge in graph.adj.get(u, []):
            weight = graph.get_edge_weight(edge, mode)
            new_dist = d + weight

            if new_dist < dist.get(v, float('inf')):
                dist[v] = new_dist
                prev_edge[v] = (u, edge)
                heapq.heappush(pq, (new_dist, v))

    end_time = time.perf_counter_ns()
    exec_us = int((end_time - start_time) / 1000)

    if not found:
        return None

    # Backtrack predecessor map to reconstruct the optimal path
    curr = target
    route_nodes = [curr]
    route_edges: List[Edge] = []

    while curr != source:
        p, edge = prev_edge[curr]
        route_edges.append(edge)
        curr = p
        route_nodes.append(curr)

    route_nodes.reverse()
    route_edges.reverse()

    tot_distance = sum(e.distance for e in route_edges)
    tot_time = sum(e.travel_time_min for e in route_edges)
    tot_fuel = sum(e.fuel_liters for e in route_edges)
    tot_fuel_cost = sum(e.fuel_cost for e in route_edges)
    tot_toll = sum(e.toll_cost for e in route_edges)
    tot_cost = dist[target]

    return RouteResult(
        source=source,
        destination=target,
        optimization=mode,
        algorithm="DIJKSTRA",
        route=route_nodes,
        distance_km=round(tot_distance, 4),
        travel_time_min=round(tot_time, 4),
        fuel_liters=round(tot_fuel, 4),
        fuel_cost=round(tot_fuel_cost, 4),
        toll_cost=round(tot_toll, 4),
        total_cost=round(tot_cost, 4),
        nodes_explored=nodes_explored,
        execution_time_us=exec_us
    )

File: rust/src/dijkstra.rs
//! Dijkstra's single-source shortest path algorithm for physical road networks.
//!
//! Implements a binary max-heap variant of Dijkstra's algorithm using Rust's
//! standard `BinaryHeap<State>` with reversed cost ordering to simulate a min-heap.
//! Stale heap entries are discarded with lazy deletion for efficiency on dense graphs.
//!
//! **Complexity**: O((V + E) log V) time, O(V) space.
//!
//! The implementation is read-only with respect to the graph, enabling safe
//! invocation across parallel threads via `Rayon` work-stealing.

use crate::graph::Graph;
use crate::models::RouteResult;
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};
use std::time::Instant;

/// Priority queue entry representing a graph vertex with its accumulated cost.
///
/// Implements `Ord` with reversed comparison so that `BinaryHeap` acts as a min-heap,
/// always yielding the vertex with the smallest accumulated cost first.
#[derive(Copy, Clone, PartialEq)]
struct State {
    cost: f64,
    node: usize,
}

impl Eq for State {}

impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        other
            .cost
            .partial_cmp(&self.cost)
            .unwrap_or(Ordering::Equal)
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

/// Computes the cost-optimal route between two vertices using Dijkstra's algorithm.
///
/// Executes a standard priority-queue-driven relaxation over the directed graph
/// adjacency structure using the pre-computed multi-objective edge costs. Stale
/// (superseded) heap entries are discarded via lazy deletion.
///
/// Path reconstruction proceeds via backtracking through the predecessor map,
/// accumulating physical metrics (distance, time, fuel, costs) along the route.
///
/// # Arguments
/// * `graph`  - Pre-built `Graph` with physics-evaluated edge costs.
/// * `source` - Origin vertex identifier.
/// * `target` - Destination vertex identifier.
/// * `mode`   - Optimization objective: `"FASTEST"`, `"CHEAPEST"`, or `"BALANCED"`.
///
/// # Returns
/// `Some(RouteResult)` with the optimal path and physical metrics,
/// or `None` if no path exists between source and target.
pub fn solve_dijkstra(graph: &Graph, source: usize, target: usize, mode: &str) -> Option<RouteResult> {
    let start_time = Instant::now();

    if !graph.nodes.contains_key(&source) || !graph.nodes.contains_key(&target) {
        return None;
    }

    if source == target {
        return Some(RouteResult {
            source,
            destination: target,
            optimization: mode.to_string(),
            algorithm: "DIJKSTRA".to_string(),
            route: vec![source],
            distance_km: 0.0,
            travel_time_min: 0.0,
            fuel_liters: 0.0,
            fuel_cost: 0.0,
            toll_cost: 0.0,
            total_cost: 0.0,
            nodes_explored: 1,
            execution_time_us: start_time.elapsed().as_micros() as u64,
        });
    }

    // dist[v] = minimum accumulated cost recorded to vertex v
    let mut dist: HashMap<usize, f64> = HashMap::new();
    // prev[v] = (predecessor_vertex, edge_index) for path reconstruction
    let mut prev: HashMap<usize, (usize, usize)> = HashMap::new();
    let mut heap: BinaryHeap<State> = BinaryHeap::new();

    dist.insert(source, 0.0);
    heap.push(State { cost: 0.0, node: source });

    let mut nodes_explored = 0;
    let mut found = false;

    while let Some(State { cost, node }) = heap.pop() {
        if let Some(&d) = dist.get(&node) {
            if cost > d {
                continue;
            }
        }

        nodes_explored += 1;

        if node == target {
            found = true;
            break;
        }

        if let Some(neighbors) = graph.adj.get(&node) {
            for &(next_node, edge_idx) in neighbors {
                let edge = &graph.edges[edge_idx];
                let weight = graph.get_edge_weight(edge, mode);
                let next_cost = cost + weight;

                let is_shorter = match dist.get(&next_node) {
                    Some(&curr_dist) => next_cost < curr_dist,
                    None => true,
                };

                if is_shorter {
                    dist.insert(next_node, next_cost);
                    prev.insert(next_node, (node, edge_idx));
                    heap.push(State {
                        cost: next_cost,
                        node: next_node,
                    });
                }
            }
        }
    }

    let exec_us = start_time.elapsed().as_micros() as u64;

    if !found {
        return None;
    }

    // Backtrack predecessor map to reconstruct the optimal node sequence
    let mut route_nodes = Vec::new();
    let mut route_edges = Vec::new();
    let mut curr = target;
    route_nodes.push(curr);

    while curr != source {
        let (p, edge_idx) = prev[&curr];
        route_edges.push(edge_idx);
        curr = p;
        route_nodes.push(curr);
    }

    route_nodes.reverse();
    route_edges.reverse();

    let mut tot_dist = 0.0;
    let mut tot_time = 0.0;
    let mut tot_fuel = 0.0;
    let mut tot_fuel_cost = 0.0;
    let mut tot_toll = 0.0;

    for &idx in &route_edges {
        let e = &graph.edges[idx];
        tot_dist += e.distance;
        tot_time += e.travel_time_min;
        tot_fuel += e.fuel_liters;
        tot_fuel_cost += e.fuel_cost;
        tot_toll += e.toll_cost;
    }

    let total_cost = dist[&target];

    Some(RouteResult {
        source,
        destination: target,
        optimization: mode.to_string(),
        algorithm: "DIJKSTRA".to_string(),
        route: route_nodes,
        distance_km: (tot_dist * 10000.0).round() / 10000.0,
        travel_time_min: (tot_time * 10000.0).round() / 10000.0,
        fuel_liters: (tot_fuel * 10000.0).round() / 10000.0,
        fuel_cost: (tot_fuel_cost * 10000.0).round() / 10000.0,
        toll_cost: (tot_toll * 10000.0).round() / 10000.0,
        total_cost: (total_cost * 10000.0).round() / 10000.0,
        nodes_explored,
        execution_time_us: exec_us,
    })
}
